rust_src/src/dired_unix.rs

```diff
@@ -95,6 +100,270 @@ impl LispObjectExt for LispObject {
+            attrs,
+            id_format,
+            files: Vec::new(),
+            files_attrs: Vec::new(),
+        }
+    }
+
+    // read_dir() does not return .&.. so bolt them on
+    fn add_dots(&mut self) {
+        let d = String::from(".");
+        let dd = String::from("..");
+        let mut df = d.clone(); // dot final name
+        let mut ddf = dd.clone();
+        if self.full {
+            df = self.get_full(d.to_owned());
+            ddf = self.get_full(dd.to_owned());
+        }
+
+        if self.match_re.is_nil() {
+            self.files.push(df.clone());
+            self.files.push(ddf.clone());
+        } else {
+            let re = RegEx::new(self.match_re);
+
+            // always match dot(s) !fullpath
+            if re.is_match(d.as_str()) {
+                self.files.push(df.clone());
+            }
+            if re.is_match(dd.as_str()) {
+                self.files.push(ddf.clone());
+            }
+        }
+    }
+
+    fn get(&mut self) -> io::Result<()> {
+        let dir = self.directory.clone();
+        let dir_p = Path::new(&dir);
+
+        if !dir_p.is_dir() {
+            xsignal!(
+                Qfile_missing,
+                LispObject::from("Opening directory: no such file or directory"),
+                self.directory.to_bstring()
+            );
+        }
+
+        self.add_dots();
+
+        for file in fs::read_dir(dir_p)? {
+            let file = file?;
+            let f_enc = file.file_name().into_string().unwrap();
+            let f_enc_lo = LispObject::from(f_enc.as_str()); // encoded
+            let f_dec_lo = unsafe { decode_file_name(f_enc_lo) }; // decoded
+            let f = f_dec_lo.to_stdstring();
+
+            let mut ff = f.clone();
+            if self.full {
+                ff = self.get_full(f.to_owned());
+            }
+
+            if self.match_re.is_nil() {
+                self.files.push(ff.clone());
+            } else {
+                let re = RegEx::new(self.match_re);
+                if re.is_match(f.as_str()) {
+                    self.files.push(ff.clone());
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn get_attrs(&mut self) {
+        for f in &self.files {
+            let mut ff = f.clone();
+            // if no full path in files make it full for file_attributes_core
+            if !self.full {
+                ff = self.get_full(f.to_owned());
+            }
+            let fattrs = file_attributes_core(LispObject::from(ff.as_str()), self.id_format);
+            self.files_attrs.push(fattrs);
+        }
+    }
+
+    // Assumes expand_file_name haz already run on
+    // self.directory and it will drop >1 trailing '/'.
+    // So here make sure just one '/' between dir&file.
+    fn get_full(&self, file: String) -> String {
+        let last_char = self.directory.as_str().chars().last().unwrap();
+        if last_char == '/' {
+            self.directory.clone() + &file.clone()
```

Comment:
`&file.clone()` is definitely unneeded, as the `Add` trait impl copies from its right-hand argument into its left-hand argument. See https://doc.rust-lang.org/std/string/struct.String.html#impl-Add%3C%26%27a%20str%3E for details.

Reply:
Agreed -- I got a little clone happy -- I'll cleanup all unnecessary clones.